modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb

```diff
@@ -0,0 +1,135 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  module Peripherals
+    module ConnectionValidators
+      module Nextcloud
+        class AuthenticationValidator < BaseValidator
+          def initialize(storage)
+            super
+            @user = User.current
+          end
+
+          def call
+            catch :interrupted do
+              @storage.authenticate_via_idp? ? validate_sso : validate_oauth
+            end
+
+            @results
+          end
+
+          private
+
+          def validate_oauth
+            @results = {
+              existing_token: CheckResult.skipped(:existing_token),
+              user_bound_request: CheckResult.skipped(:user_bound_request)
+            }
+
+            oauth_token
+            user_bound_request
+          end
+
+          def oauth_token
+            if OAuthClientToken.where(user: @user, oauth_client: @storage.oauth_client).any?
+              pass_check(:existing_token)
+            else
+              warn_check(:existing_token, message(:oauth_token_missing))
```

Comment:
❓ why is it `warn_check` and not a `fail_check`?

Reply:
same in `non_provisioned_user` and `non_oidc_provisioned_user`

Reply:
I guess because the check doesn't quite mean that stuff is entirely broken, but rather "I can't check further, that makes me sad". So we need to abort, but we can't say for sure that it's broken.

Reply:
Yep, it is not a failure as I can't simply get a token without going through the entire OAuth2 process.

This makes the context slightly different. =/